daphne_worker/src/dap.rs

```diff
@@ -64,29 +65,90 @@ pub(crate) fn dap_response_to_worker(resp: DapResponse) -> Result<Response> {
     Ok(worker_resp)
 }
 
+#[async_trait(?Send)]
 impl<D> HpkeDecrypter for DaphneWorkerConfig<D> {
-    fn get_hpke_config_for(&self, _task_id: &Id) -> Option<&HpkeConfig> {
-        if self.hpke_receiver_config_list.is_empty() {
-            return None;
-        }
+    async fn get_hpke_config_for(
+        &self,
+        _task_id: &Id,
+    ) -> std::result::Result<Option<HpkeConfig>, DapError> {
+        let kv_store = self.kv("DAP_HPKE_RECEIVER_CONFIG_STORE").map_err(dap_err)?;
+        let keys = kv_store
+            .list()
+            .execute()
+            .await
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
 
-        // Always advertise the first HPKE config in the list.
-        Some(&(self.hpke_receiver_config_list[0]).config)
+        if keys.keys.is_empty() {
+            // Generate a new random HPKE config ID.
+            let hpke_config_id: u8 = rand::thread_rng().gen();
+            let name = format!("hpke_receiver_config/{}", hpke_config_id);
+
+            // Generate a new HPKE receiver config.
+            //
+            // NOTE(nakatsuka-y): The following logic should be changed when we support
+            // multiple HPKE KEMs in the future.
+            if self.global_config.supported_hpke_kems.len() != 1 {
+                return Err(DapError::Fatal(
+                    "The number of supported HPKE KEMs must be 1".to_string(),
+                ));
+            }
+            let hpke_receiver_config =
+                HpkeReceiverConfig::gen(hpke_config_id, self.global_config.supported_hpke_kems[0]);
+
+            // Store newly generated HPKE receiver config into KV.
+            kv_store
+                .put(&name, hpke_receiver_config.clone())
+                .map_err(|e| DapError::Fatal(e.to_string()))?
+                .execute()
+                .await
+                .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+            // Add HPKE receiver config in list.
+            Ok(Some(hpke_receiver_config.config))
+        } else {
+            // Return the first HPKE receiver config.
+            let builder = kv_store.get(&keys.keys[0].name);
+            if let Some(hpke_receiver_config) = builder
+                .json::<HpkeReceiverConfig>()
+                .await
+                .map_err(|e| DapError::Fatal(e.to_string()))?
+            {
+                Ok(Some(hpke_receiver_config.config))
+            } else {
+                Ok(None)
+            }
+        }
     }
 
-    fn can_hpke_decrypt(&self, _task_id: &Id, config_id: u8) -> bool {
-        self.get_hpke_receiver_config_for(config_id).is_some()
+    // This is called by the Leader when handling the `/upload` request.
+    // Get HPKE receiver config from KV.
```

Comment:
Function block comments have the format
```rust
///
/// 
```

*But*, this comment isn't super helpful, since this a trait method and the functionality of the method is documented by the trait. If you think additional context would improve readability, consider adding a comment in the function itself.
```suggestion
```

Reply:
Ditto :)